lib/src/native/realm_core.dart

```diff
@@ -500,52 +502,45 @@ class _RealmCore {
 
     if (syncError is ClientResetError) {
       syncConfig.clientResetHandler.onManualReset?.call(syncError);
-    } else {
-      syncConfig.syncErrorHandler(syncError);
+      return;
+    }
+
+    syncConfig.syncErrorHandler(syncError);
+  }
+
+  static void _callResetCallback(FutureOr<void> Function() callback, Pointer<Void> unlockCallbackFunc) async {
+    bool success = true;
+    try {
+      await callback();
+    } catch (error) {
+      success = false;
+      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
+    } finally {
+      _realmLib.realm_dart_invoke_unlock_callback(success, unlockCallbackFunc);
     }
   }
 
   static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
-    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset;
-    if (beforeResetCallback == null) {
```

Comment:
The callback is only registered if the dart callback was provided. So if it is invoked then we are sure this can't be null.